cassiopeia/simulator/Cas9LineageTracingDataSimulator.py

```diff
@@ -0,0 +1,327 @@
+                raise DataSimulatorError(
+                    "Mutation rate needs to be" " non-negative."
+                )
+            self.mutation_rate_per_character = mutation_rate
+
+        self.mutation_priors = mutation_priors
+        if self.mutation_priors is None:
+            self.mutation_priors = {}
+            probabilites = [
+                state_distribution() for _ in range(number_of_states)
+            ]
+            Z = np.sum(probabilites)
+            for i in range(number_of_states):
+                self.mutation_priors[i + 1] = probabilites[i] / Z
+        else:
+            Z = np.sum([v for v in self.mutation_priors.values()])
+            if not math.isclose(Z, 1.0):
+                raise DataSimulatorError("Mutation priors do not sum to 1.")
+
+        self.heritable_silencing_rate = heritable_silencing_rate
+        self.stochastic_silencing_rate = stochastic_silencing_rate
+
+        self.random_seed = random_seed
+
+    def overlay_data(self, tree: CassiopeiaTree):
+        """Overlays Cas9-based lineage tracing data onto the CassiopeiaTree.
+
+        Args:
+            tree: Input CassiopeiaTree
+        """
+
+        if self.random_seed is not None:
+            np.random.seed(self.random_seed)
+
+        number_of_characters = self.number_of_cassettes * self.size_of_cassette
+
+        # initialize character states
+        character_matrix = {}
+        for node in tree.nodes:
+            character_matrix[node] = [-1] * number_of_characters
+
+        for node in tree.depth_first_traverse_nodes(tree.root, postorder=False):
+
+            if tree.is_root(node):
+                character_matrix[node] = [0] * number_of_characters
+                continue
+
+            parent = tree.parent(node)
+            t = tree.get_time(node) - tree.get_time(parent)
+
+            character_array = character_matrix[parent]
+            open_sites = [
+                c
+                for c in range(len(character_array))
+                if character_array[c] == 0
+            ]
+
+            new_cuts = []
+            for site in open_sites:
+                mutation_rate = self.mutation_rate_per_character[site]
+                p = 1 - (np.exp(-t * mutation_rate))
+
+                if np.random.uniform() < p:
+                    new_cuts.append(site)
+
+            # collapse cuts that are on the same cassette
+            cuts_remaining = new_cuts
+            if self.size_of_cassette > 1:
+                character_array, cuts_remaining = self.collapse_sites(
+                    character_array, new_cuts
+                )
+
+            # introduce new states at cut sites
+            character_array = self.introduce_states(
+                character_array, cuts_remaining
+            )
+
+            # silence cassettes
+            character_array = self.silence_cassettes(
+                character_array, self.heritable_silencing_rate
+            )
```

Comment:
Is it true that silencing only happens at division events?

Reply:
Yes and no. It might as well just happen at division events in the simulation -- because if the silencing happens at all during a cell's lifetime it's going to get passed on to its children. But more generally, we don't really know when the heritable silencing happens.

Reply:
I guess what changes is our interpretation of the `heritable_silencing_rate` parameter. The way it's implemented now, it's dependent on the underlying tree topology (how many generations occur determines the probability that a silencing event will happen in any given character in any given cell). Whereas if we used the silencing rate as a true rate in an exponential process like we're doing for the mutation rate, it might be easier to build estimators for. (If a silencing event happens at all on a branch, then it'll just override any other mutation that might have occurred at a site on that branch in the case where we use an exponential).

Reply:
Nice catch @richardyz98 . I agree with @richardyz98 , the `heritable_silencing_rate` is per division event and not per unit time. @mattjones315 which is more reasonable biologically? We should implement the more biologically reasonable one. I feel like per time unit (as @richardyz98 mentioned) is more reasonable: the longer the cell lives, the more likely a heritable dropout is.

Reply:
This is related to a point I made above about the intersection between biology and the simulation -- I'm actually not entirely sure when the transcriptional silencing would happen (i.e., if it's at a division event or if it could occur during a cell's lifetime). I agree though, for the purposes of building estimators, let's encode this as a time-dependent process. I'm happy to add this to the simulator!

Reply:
Gotcha, then I'm fine either way! Let's just remember to document it.

Reply:
In my latest commit I've updated the tree simulator to have a time-dependence to the transcriptional silencing. here, I think it makes sense to model -- especially because we'll want to use some estimators here